File: mediaworker/src/utils/rbac.py
"""Проверка иерархических прав (копия минимальной логики billing RBAC).

Домены разделены: mediaworker не импортирует код billing, поэтому крошечная
проверка ``has_perm`` продублирована здесь.
"""

from __future__ import annotations


def has_perm(perms: dict | None, path: str) -> bool:
    """Проверить, даёт ли набор прав ``perms`` доступ к ``path`` (с наследованием)."""
    if not perms:
        return False
    node: object = perms
    for seg in path.split("."):
        if node is True:
            return True
        if not isinstance(node, dict):
            return False
        if node.get("*") is True:
            return True
        if seg not in node:
            return False
        node = node[seg]
    return node is True or isinstance(node, dict)


__all__ = ["has_perm"]

File: adminui/src/lib/rbac.ts
/**
 * Клиентское зеркало `security/rbac.py::has_perm` (биллинг, Python).
 *
 * Держать в точном соответствии с backend-реализацией — это только для UI-гейтинга
 * (скрыть пункт меню/кнопку), окончательную проверку всегда делает backend.
 */
export type PermNode = true | { [key: string]: PermNode } | Record<string, never>

export function hasPerm(perms: PermNode | null | undefined, path: string): boolean {
  if (!perms) return false
  let node: PermNode = perms
  for (const seg of path.split(".")) {
    if (node === true) return true
    if (typeof node !== "object") return false
    if (node["*"] === true) return true
    if (!(seg in node)) return false
    node = node[seg]
  }
  return node === true || typeof node === "object"
}
